### PR TITLE
Add passkey testing support

### PR DESCRIPTION
**Pull request by ikerexxe**
The following packages, capabilities and binaries are needed to perform the passkey testing in an automated way.

### CODE REVIEW

**Comment by ikerexxe** [comment]
@madhuriupadhye, @spoore1 these are the changes needed to run the passkey testing in an automated way. It allows to playback the recorded umockdev data, but not to record it. For that purpose additional changes are needed and I don't think they should be included in this repository/branch. If you still need to record anything please use https://github.com/ikerexxe/sssd-ci-containers/tree/passkey.

File: docker-compose.yml [hunk 1]
```diff
@@ -91,6 +92,7 @@ services:
     - NET_ADMIN
     - AUDIT_WRITE
     - SYS_CHROOT
+    #- SYS_RAWIO # Only needed if you are using real FIDO2 tokens
```
Comment (pbrezina) [inline docker-compose.yml:95]: This should be in different commit and perhaps use "only needed if you require access to real devices like FIDO2 tokens"?

File: src/ansible/roles/packages/tasks/Fedora.yml [hunk 2]
```diff
@@ -74,6 +74,15 @@
...
         libtevent
+  - name: Install packages required for passkey testing
+    dnf:
+      state: present
+      name:
+        - gcc
+        - openssl-devel
+        - umockdev
+    when: (ansible_facts['distribution'] == "CentOS" and ansible_distribution_major_version | int >= 9) or
+          (ansible_facts['distribution'] == "Fedora")
```
Comment (pbrezina) [inline src/ansible/roles/packages/tasks/Fedora.yml:85]: It would be better to add new fact to https://github.com/SSSD/sssd-ci-containers/tree/master/src/ansible/roles/facts/tasks

E.g.

```yaml
passkey_support: Yes|No
```

End then use just `when: passkey_support`

File: src/ansible/roles/client/tasks/main.yml [hunk 3]
```diff
@@ -82,3 +82,22 @@
...
     fi
+
+- name: Copy random.c to /opt
+  copy:
+    src: '{{ playbook_dir }}/../tools/random.c'
+    dest: /opt/random.c
+    owner: root
+    group: root
+    mode: 0644
+  when: (ansible_facts['distribution'] == "Fedora") or
```
Comment (pbrezina) [inline src/ansible/roles/client/tasks/main.yml:93]: Use new fact as well.

File: src/ansible/roles/client/tasks/main.yml [hunk 4]
```diff
@@ -82,3 +82,22 @@
...
+        (ansible_facts['distribution'] == "Ubuntu") or
+        (ansible_facts['distribution'] == "CentOS" and ansible_distribution_major_version | int >= 9)
+        
+- name: Build random.so needed for passkey testing
+  shell: |
+    cd /opt
+    gcc -fPIC -shared -o random.so random.c -lcrypto
+  when: (ansible_facts['distribution'] == "Fedora") or
```
Comment (pbrezina) [inline src/ansible/roles/client/tasks/main.yml:101]: New fact as well.

File: src/ansible/roles/client/tasks/main.yml (hunk 3 above)
Comment (pbrezina) [reply src/ansible/roles/client/tasks/main.yml:93]: The file should be removed after compilation.

You can also wrap it with `block` together with compilation.

**Review by pbrezina — CHANGES_REQUESTED** [review]
Thank you, see minor comments inside.

**Comment by pbrezina** [comment]
> @madhuriupadhye, @spoore1 these are the changes needed to run the passkey testing in an automated way. It allows to playback the recorded umockdev data, but not to record it. For that purpose additional changes are needed and I don't think they should be included in this repository/branch. If you still need to record anything please use https://github.com/ikerexxe/sssd-ci-containers/tree/passkey.

What additional changes are needed? Maybe we can include the as well so we can easily create new recordings in the future on newer containers.

**Comment by ikerexxe** [comment]
We would need to add the `SYS_RAWIO` capability, and find a way to export the exact device where the fido2 token is connected (the exact line changes are linked below). I don't know if those changes are compatible with a cloud environment as they may be used to access low-level stuff.

Changes:
https://github.com/ikerexxe/sssd-ci-containers/blob/passkey/Makefile#L1
https://github.com/ikerexxe/sssd-ci-containers/blob/passkey/Makefile#L12
https://github.com/ikerexxe/sssd-ci-containers/blob/passkey/docker-compose.yml#L35

File: src/ansible/roles/packages/tasks/Fedora.yml (hunk 2 above)
Comment (ikerexxe) [reply src/ansible/roles/packages/tasks/Fedora.yml]: I hope I did it correctly. Will need to check the results later.

By the way, thanks for the lesson about facts ;)

File: src/ansible/roles/client/tasks/main.yml (hunk 3 above)
Comment (ikerexxe) [reply src/ansible/roles/client/tasks/main.yml]: Done both of the improvements

File: src/ansible/roles/packages/tasks/Fedora.yml [hunk 5]
```diff
@@ -74,6 +74,14 @@
...
         libtalloc               \
         libtevent
+  - name: Install packages required for passkey testing
+    dnf:
+      state: present
+      name:
+        - gcc
+        - openssl-devel
+        - umockdev
+    when: passkey_support
```
Comment (pbrezina) [inline src/ansible/roles/packages/tasks/Fedora.yml:84]: Please align the list to the key, ie.

```
key:
- value
- value
```

File: src/ansible/roles/client/tasks/main.yml [hunk 6]
```diff
@@ -82,3 +82,24 @@
     else
       exit 0
     fi
+
+- name: Passkey testing
+  block:
+  - name: Copy random.c to /opt
+    copy:
+      src: '{{ playbook_dir }}/../tools/random.c'
```
Comment (pbrezina) [inline src/ansible/roles/client/tasks/main.yml:90]: You can put this to files folder in this role.

**Review by pbrezina — CHANGES_REQUESTED** [review]
Thank you, two more nitpicks and we can merge.

File: src/ansible/roles/packages/tasks/Ubuntu.yml [hunk 7]
```diff
@@ -49,6 +49,13 @@
       - realmd
       - sssd
       - sssd-*
+  - name: Install packages required for passkey testing
+    apt:
+      state: present
+      name:
+      - gcc
+      - libssl-dev
+      - umockdev
```
Comment (pbrezina) [inline src/ansible/roles/packages/tasks/Ubuntu.yml:58]: when: passkey_support to be consistent with Fedora?

File: docker-compose.yml [hunk 8]
```diff
@@ -91,6 +92,7 @@ services:
     - NET_ADMIN
     - AUDIT_WRITE
     - SYS_CHROOT
+    #- SYS_RAWIO # Only needed if you require access to real devices like FIDO2 tokens
```
Comment (pbrezina) [inline docker-compose.yml:95]: If this can not work on its own, I think we can remove this patch.

**Review by pbrezina — CHANGES_REQUESTED** [review]

**Comment by ikerexxe** [comment]
I added the override and the new target in the Makefile

File: src/ansible/roles/packages/tasks/Ubuntu.yml (hunk 7 above)
Comment (ikerexxe) [reply src/ansible/roles/packages/tasks/Ubuntu.yml:58]: Done

File: docker-compose.yml (hunk 8 above)
Comment (ikerexxe) [reply docker-compose.yml]: I removed this patch

File: src/ansible/roles/packages/tasks/Ubuntu.yml [hunk 9]
```diff
@@ -132,8 +140,6 @@
       - libselinux1-dev
       - libsemanage-dev
       - libsmbclient-dev
-      - libssl-dev
-      - libssl-dev
```
Comment (pbrezina) [inline src/ansible/roles/packages/tasks/Ubuntu.yml:136]: OpenSSL devel is required to build SSSD. This should stay, since it is installed only conditionally for passkey.

File: Makefile [hunk 10]
```diff
@@ -1,3 +1,5 @@
+SHELL_OUT = $(shell fido2-token -L|cut -f1 -d:)
```
Comment (pbrezina) [inline Makefile:1]: Can this be moved to up-passkey?

File: src/ansible/roles/packages/tasks/Ubuntu.yml (hunk 9 above)
Comment (ikerexxe) [reply src/ansible/roles/packages/tasks/Ubuntu.yml:136]: Added it back

File: Makefile (hunk 10 above)
Comment (ikerexxe) [reply Makefile]: Done

File: src/ansible/roles/facts/tasks/CentOS.yml [hunk 11]
```diff
@@ -1,2 +1,19 @@
...
+
+- name: Define passkey for CentOS9+
+  set_fact:
+    passkey_support: Yes
+  when: (ansible_distribution_major_version | int >= 9)
+
+- name: Define passkey for CentOS8-
+  set_fact:
+    passkey_support: No
+  when: (ansible_distribution_major_version | int < 9)
```
Comment (pbrezina) [inline src/ansible/roles/facts/tasks/CentOS.yml:19]: You can keep the Fedora include and then just override what you need (see https://github.com/SSSD/sssd-ci-containers/blob/master/src/ansible/roles/facts/tasks/RedHat.yml). It may also be clearer to separate this to CentOS8.yml and CentOS9.yml.

**Comment by pbrezina** [comment]
Thank you. Ack.

**Review by pbrezina — APPROVED** [review]
